metricbeat/module/zookeeper/server/server.go

```diff
@@ -0,0 +1,249 @@
+var versionCapturer = regexp.MustCompile(`:\s(.*),`)
+var dateCapturer = regexp.MustCompile(`built on (.*)`)
+
+func init() {
+	mb.Registry.MustAddMetricSet("zookeeper", "server", New,
+		mb.WithHostParser(parse.PassThruHostParser),
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet for fetching ZooKeeper health metrics.
+type MetricSet struct {
+	mb.BaseMetricSet
+}
+
+// New creates new instance of MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	return &MetricSet{
+		BaseMetricSet: base,
+	}, nil
+}
+
+// Fetch fetches metrics from ZooKeeper by making a tcp connection to the
+// command port and sending the "srvr" command and parsing the output.
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	outputReader, err := zookeeper.RunCommand("srvr", m.Host(), m.Module().Config().Timeout)
+	if err != nil {
+		reporter.Error(errors.Wrap(err, "srvr command failed"))
+		return
+	}
+
+	metricsetFields, err := parseSrvr(outputReader)
+	if err != nil {
+		reporter.Error(err)
+		return
+	}
+
+	reporter.Event(mb.Event{
+		MetricSetFields: metricsetFields,
+	})
+}
+
+func parseSrvr(i io.Reader) (common.MapStr, error) {
+	scanner := bufio.NewScanner(i)
+
+	output := common.MapStr{}
+
+	//Get version
+	ok := scanner.Scan()
+	if !ok {
+		return nil, errors.New("no initial successful scan, aborting")
+	}
+	output.Put("version.id", versionCapturer.FindStringSubmatch(scanner.Text())[1])
+	output.Put("version.date", dateCapturer.FindStringSubmatch(scanner.Text())[1])
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, "Zxid") {
+			xid, err := parseXid(line)
+			if err != nil {
+				return nil, errors.Wrap(err, "error parsing xid line")
+			}
+			output.Put("xid", xid)
+
+			continue
+		}
+
+		if strings.Contains(line, "Latency") {
+			latency, err := parseLatencyLine(line)
+			if err != nil {
+				return nil, errors.Wrap(err, "error parsing latency values")
+			}
+			output.Put("latency", latency)
+
+			continue
+		}
+
+		if strings.Contains(line, "Proposal sizes") {
+			proposalSizes, err := parseProposalSizes(line)
+			if err != nil {
+				return nil, errors.Wrap(err, "error parsing proposal sizes line")
+			}
+			output.Put("proposal_sizes", proposalSizes)
+
+			continue
+		}
+
+		if strings.Contains(line, "Mode") {
+			output.Put("mode", strings.Split(line, " ")[1])
+			continue
+		}
+
+		// If code reaches here easy to parse lines or blank lines like the following:
+		// Received: 46
+		//
+		// Sent: 45
+		// Connections: 1
+		// Outstanding: 0
+		results := fieldsCapturer.FindAllStringSubmatch(line, -1)
+		if len(results) == 0 {
+			//probably a blank line
+			continue
+		}
+
+		for _, result := range results {
+			val, err := strconv.ParseInt(result[2], 10, 64)
+			if err != nil {
+				return nil, err
+			}
+			output.Put(strings.ToLower(strings.Replace(result[1], " ", "_", -1)), val)
+		}
+	}
+	return output, nil
+}
+
+func parseXid(line string) (common.MapStr, error){
```

Comment:
Any chance to get some unit tests for the parse functions?

Reply:
Maybe you have missed it but those functions are tested by the _global_ `parseSrvr` that uses all of them. In that test all possible values are checked so I considered it tested (75% of coverage is achieved too).

Do you want individual tests too or it is enough with the _global_ test?

Reply:
Works for me for the moment. In general I think we mainly test the "correct" cases but probably also should check the error cases. Not critical here.